# Coffee survey data - Datenverständnis & Aufbereitung mit pandas 

Data retrieved from: https://github.com/rfordatascience/tidytuesday/blob/main/data/2024/2024-05-14/coffee_survey.csv

Extended dataset available from:
https://www.kaggle.com/datasets/datalab351/great-american-coffee-taste-test/data

For overview of variables see: https://github.com/rfordatascience/tidytuesday/blob/main/data/2024/2024-05-14/readme.md

**Projektschritt 1:**
Überblick über die Variablen, fehlende Werte identifizieren und behandeln, kategorische Variablen vereinheitlichen / umkodieren, Daten bereinigen (z.B. Schreibvarianten, Ausreißer behandeln)

## Beschreibung des Datensatzes

Der bekante Youtube-Barista James Hoffmann und Cometeer haben 2023 einen Kaffee-Geschmackstest durchgeführt. Cometeer verschickte 5000 Kaffee-Kits in ganz Amerika. Die Kits enthielten vier verschiedene Kaffeesorten - vor-extrahiert und tiefgefroren. Die Tester schmolzen und verdünnten die Kaffeekapseln, um ein weitgehend identisches Geschmackserlebnis zu erhalten. Verkostung und Bewertung erfolgten blind. Nachdem die Antworten auf die Umfrage gesammelt worden waren (die Daten wurden bereitgestellt), wurden einige Eigenschaften des Kaffees aufgedeckt. 

**Kaffeesorten (aufgedeckt nach der Umfrage)**

A. Light roast, washed process

B. Medium roast

C. Dark roast

D. Light roast, natural process (fermented, funky/fruity)


**Die Kaffeesorten wurden für drei Hauptfragen ausgewählt:**
1. Röstgradpräferenz (light (A) vs. medium (B) vs. dark (C))

2. Vorliebe für Fruchtigkeit (washed (A) vs. natural/fermented (D))

3. Beliebtester Kaffee insgesamt



Beschreibung übersetzt und angepasst von: https://www.kaggle.com/datasets/datalab351/great-american-coffee-taste-test/data

## Daten laden

In [1]:
# Import modules

import pandas as pd
import numpy as np

In [2]:
# Read csv-file

df = pd.read_csv("data/coffee_survey.csv")

## Überblick verschaffen

In [3]:
# Dataframe information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4042 entries, 0 to 4041
Data columns (total 57 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   submission_id                 4042 non-null   object 
 1   age                           4011 non-null   object 
 2   cups                          3949 non-null   object 
 3   where_drink                   3972 non-null   object 
 4   brew                          3657 non-null   object 
 5   brew_other                    678 non-null    object 
 6   purchase                      710 non-null    object 
 7   purchase_other                31 non-null     object 
 8   favorite                      3980 non-null   object 
 9   favorite_specify              114 non-null    object 
 10  additions                     3959 non-null   object 
 11  additions_other               48 non-null     object 
 12  dairy                         1686 non-null   object 
 13  swe

In [4]:
df.head()

submission_id              age cups where_drink  \
0        gMR29l  18-24 years old  NaN         NaN   
1        BkPN0e  25-34 years old  NaN         NaN   
2        W5G8jj  25-34 years old  NaN         NaN   
3        4xWgGr  35-44 years old  NaN         NaN   
4        QD27Q8  25-34 years old  NaN         NaN   

                                          brew brew_other purchase  \
0                                          NaN        NaN      NaN   
1  Pod/capsule machine (e.g. Keurig/Nespresso)        NaN      NaN   
2                          Bean-to-cup machine        NaN      NaN   
3     Coffee brewing machine (e.g. Mr. Coffee)        NaN      NaN   
4                                    Pour over        NaN      NaN   

  purchase_other             favorite favorite_specify  ... spent_equipment  \
0            NaN  Regular drip coffee              NaN  ...             NaN   
1            NaN          Iced coffee              NaN  ...             NaN   
2            NaN  Regular drip coffee              NaN  ...             NaN   
3            NaN          Iced coffee              NaN  ...             NaN   
4            NaN                Latte              NaN  ...             NaN   

  value_equipment gender gender_specify education_level ethnicity_race  \
0             NaN    NaN            NaN             NaN            NaN   
1             NaN    NaN            NaN             NaN            NaN   
2             NaN    NaN            NaN             NaN            NaN   
3             NaN    NaN            NaN             NaN            NaN   
4             NaN    NaN            NaN             NaN            NaN   

  ethnicity_race_specify employment_status  number_children  \
0                    NaN               NaN              NaN   
1                    NaN               NaN              NaN   
2                    NaN               NaN              NaN   
3                    NaN               NaN              NaN   
4                    NaN               NaN              NaN   

   political_affiliation  
0                    NaN  
1                    NaN  
2                    NaN  
3                    NaN  
4                    NaN  

[5 rows x 57 columns]

### Erste Beobachtungen Datensatz

- es gibt 57 Variablen
- der Datensatz hat 4042 Zeilen
- es gibt numerische und kategorische Variablen
- es gibt überall missing values
- für einige Variablen gibt es mehr missing values als existierende values, das sind aber oft "other"-Variablen, die dann vermutlich nicht zutreffend sind
- was ist wfh? -> work from home
- es gibt nicht so viele numerische Variablen

In [5]:
# Variablen genauer untersuchen

for col in df.columns:
    print(f"no. of unique values for {col}:", df[col].nunique())

no. of unique values for submission_id: 4042
no. of unique values for age: 7
no. of unique values for cups: 6
no. of unique values for where_drink: 65
no. of unique values for brew: 449
no. of unique values for brew_other: 160
no. of unique values for purchase: 89
no. of unique values for purchase_other: 26
no. of unique values for favorite: 12
no. of unique values for favorite_specify: 77
no. of unique values for additions: 53
no. of unique values for additions_other: 42
no. of unique values for dairy: 175
no. of unique values for sweetener: 82
no. of unique values for style: 12
no. of unique values for strength: 5
no. of unique values for roast_level: 7
no. of unique values for caffeine: 3
no. of unique values for expertise: 10
no. of unique values for coffee_a_bitterness: 5
no. of unique values for coffee_a_acidity: 5
no. of unique values for coffee_a_personal_preference: 5
no. of unique values for coffee_a_notes: 2317
no. of unique values for coffee_b_bitterness: 5
no. of unique va

In [6]:
# Beispielhaft einzelne Variablen anschauen

print(df["age"].unique())
print(df["cups"].unique())
print(df["where_drink"].value_counts())
print(df["brew"].value_counts())
print(df["coffee_a_acidity"].unique())
print(df["number_children"].unique())

['18-24 years old' '25-34 years old' '35-44 years old' '55-64 years old'
 nan '<18 years old' '>65 years old' '45-54 years old']
[nan 'Less than 1' '2' '1' '3' 'More than 4' '4']
where_drink
At home                                         1625
At home, At a cafe                               416
At home, At the office                           407
At the office, At home                           229
At the office                                    141
                                                ... 
At a cafe, On the go, At the office                1
On the go, At a cafe, At the office                1
None of these, At a cafe                           1
On the go, At the office, At home, At a cafe       1
At a cafe, On the go                               1
Name: count, Length: 65, dtype: int64
brew
Pour over                                                                                 575
Pour over, Espresso                                                                      

### Erste Beobachtungen Variablen
- einige Variablen haben eine große Anzahl "unique Values", was darauf hindeuten könnte, dass es keine festen Antwortmöglichkeiten gab, sondern freier Text eingegeben wurde oder mehrere Antwortmöglichkeiten ausgewählt werden konnten
- Level der kategorischen Variablen muss definiert werden
- was tun mit Variablen, bei denen es unterschiedliche Kombinationen von Antwortmöglichkeiten gibt?
    - neue Variable, bei der Antworten und häufige Kombinationen bleiben und seltene Kombinationen zusammengefasst werden unter Bezeichung wie "mehrere/mehr als..."
    - one-hot encoding
- Duplikate scheint es nicht zu geben, da die Anzahl "unique values" bei der Variable "submission_id" der Anzahl Zeilen in den Daten entspricht
- Number of children: es gibt keine Antwort "0" oder "keine" -> war die Umfrage vielleicht schlecht designed, dass es diese Antwortmöglichkeit nicht gab und wir nans jetzt nicht unterscheiden können von "keine Kinder"?

In [7]:
# missing values
print(df.isna().any()) #  tatsächlich für alle Variablen missings
print(df["age"].isna().sum()) # Anzahl missings für einzelne Variable zählen

submission_id                   False
age                              True
cups                             True
where_drink                      True
brew                             True
brew_other                       True
purchase                         True
purchase_other                   True
favorite                         True
favorite_specify                 True
additions                        True
additions_other                  True
dairy                            True
sweetener                        True
style                            True
strength                         True
roast_level                      True
caffeine                         True
expertise                        True
coffee_a_bitterness              True
coffee_a_acidity                 True
coffee_a_personal_preference     True
coffee_a_notes                   True
coffee_b_bitterness              True
coffee_b_acidity                 True
coffee_b_personal_preference     True
coffee_b_not

In [8]:
# descriptives of numeric variables

df.describe()

expertise  coffee_a_bitterness  coffee_a_acidity  \
count  3938.000000          3798.000000       3779.000000   
mean      5.693499             2.141127          3.634824   
std       1.948867             0.947163          0.982113   
min       1.000000             1.000000          1.000000   
25%       5.000000             1.000000          3.000000   
50%       6.000000             2.000000          4.000000   
75%       7.000000             3.000000          4.000000   
max      10.000000             5.000000          5.000000   

       coffee_a_personal_preference  coffee_b_bitterness  coffee_b_acidity  \
count                   3789.000000          3780.000000       3767.000000   
mean                       3.310900             3.013228          2.223786   
std                        1.185953             0.992875          0.865389   
min                        1.000000             1.000000          1.000000   
25%                        2.000000             2.000000          2.000000   
50%                        3.000000             3.000000          2.000000   
75%                        4.000000             4.000000          3.000000   
max                        5.000000             5.000000          5.000000   

       coffee_b_personal_preference  coffee_c_bitterness  coffee_c_acidity  \
count                   3773.000000          3764.000000       3751.000000   
mean                       3.068646             3.071998          2.366836   
std                        1.113546             0.999267          0.921048   
min                        1.000000             1.000000          1.000000   
25%                        2.000000             2.000000          2.000000   
50%                        3.000000             3.000000          2.000000   
75%                        4.000000             4.000000          3.000000   
max                        5.000000             5.000000          5.000000   

       coffee_c_personal_preference  coffee_d_bitterness  coffee_d_acidity  \
count                   3766.000000          3767.000000       3765.000000   
mean                       3.064790             2.162729          3.858167   
std                        1.128431             1.081546          1.007973   
min                        1.000000             1.000000          1.000000   
25%                        2.000000             1.000000          3.000000   
50%                        3.000000             2.000000          4.000000   
75%                        4.000000             3.000000          5.000000   
max                        5.000000             5.000000          5.000000   

       coffee_d_personal_preference  
count                   3764.000000  
mean                       3.375930  
std                        1.452504  
min                        1.000000  
25%                        2.000000  
50%                        4.000000  
75%                        5.000000  
max                        5.000000

In [9]:
# Descriptives of categorical variables

df.describe(include=["object"])

submission_id              age  cups where_drink       brew brew_other  \
count           4042             4011  3949        3972       3657        678   
unique          4042                7     6          65        449        160   
top           gMR29l  25-34 years old     2     At home  Pour over  Aeropress   
freq               1             1986  1663        1625        575        335   

                     purchase purchase_other  favorite favorite_specify  ...  \
count                     710             31      3980              114  ...   
unique                     89             26        12               77  ...   
top     Specialty coffee shop       Cometeer  Pourover       Flat White  ...   
freq                      116              2      1084                9  ...   

         spent_equipment value_equipment gender      gender_specify  \
count               3506            3494   3523                  12   
unique                 7               2      5                  11   
top     More than $1,000             Yes   Male  Transgender Female   
freq                 780            3318   2524                   2   

          education_level   ethnicity_race ethnicity_race_specify  \
count                3438             3418                    105   
unique                  6                6                     82   
top     Bachelor's degree  White/Caucasian                  Mixed   
freq                 1759             2626                      7   

         employment_status number_children political_affiliation  
count                 3419             856                  3289  
unique                   6               4                     4  
top     Employed full-time               2              Democrat  
freq                  2708             402                  1768  

[4 rows x 44 columns]

## Datensatz aufräumen

### Kategorische Variablen mit Einfachantwort

In [10]:
# converting the variables of dtype "object":
# those with only a small number of different answers (< 13) probably had fixed answer possibilities and are therefore transformed to dtype "category"
# the cut-off at 12 was determined by looking at nunique for all variables above and looking at the unique values for some 
# (only problem might be gender_specify because this seems to be free text but also has 12 unique values)

df_obj = df.select_dtypes(include='object')  # nur object-Spalten

for col in df_obj.columns:
    if (df[col].nunique() < 13) | (col == "submission_id"):
        df[col] = df[col].astype("category")

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4042 entries, 0 to 4041
Data columns (total 57 columns):
 #   Column                        Non-Null Count  Dtype   
---  ------                        --------------  -----   
 0   submission_id                 4042 non-null   category
 1   age                           4011 non-null   category
 2   cups                          3949 non-null   category
 3   where_drink                   3972 non-null   object  
 4   brew                          3657 non-null   object  
 5   brew_other                    678 non-null    object  
 6   purchase                      710 non-null    object  
 7   purchase_other                31 non-null     object  
 8   favorite                      3980 non-null   category
 9   favorite_specify              114 non-null    object  
 10  additions                     3959 non-null   object  
 11  additions_other               48 non-null     object  
 12  dairy                         1686 non-null   ob

In [11]:
df.loc[:,"coffee_a_bitterness":"coffee_d_notes"].apply(lambda col: col.unique())
# coffees seem to be rated on Likert scale from 1-5 for bitterness, acidity, and personal preferences. Notes for each seems to contain free text


coffee_a_bitterness                                [nan, 4.0, 1.0, 3.0, 2.0, 5.0]
coffee_a_acidity                                   [nan, 4.0, 1.0, 3.0, 5.0, 2.0]
coffee_a_personal_preference                       [nan, 4.0, 1.0, 3.0, 2.0, 5.0]
coffee_a_notes                  [nan, a, Bitter, soft, This is just an audit t...
coffee_b_bitterness                                [nan, 4.0, 1.0, 3.0, 2.0, 5.0]
coffee_b_acidity                                   [nan, 4.0, 1.0, 3.0, 2.0, 5.0]
coffee_b_personal_preference                       [nan, 4.0, 1.0, 3.0, 5.0, 2.0]
coffee_b_notes                  [nan, b, Hazelnut, bold, I personally did not ...
coffee_c_bitterness                                [nan, 4.0, 1.0, 3.0, 2.0, 5.0]
coffee_c_acidity                                   [nan, 4.0, 1.0, 3.0, 2.0, 5.0]
coffee_c_personal_preference                       [nan, 4.0, 1.0, 3.0, 2.0, 5.0]
coffee_c_notes                  [nan, c, Dark roast, But the form does not ask...
coffee_d_bittern

In [12]:
df["expertise"].unique()
# expertise seems to be rated on a Likert scale from 1-10

array([nan,  7., 10.,  6.,  4.,  9.,  8.,  5.,  2.,  1.,  3.])

In [15]:
# float-Variablen umwandeln in Kategorien (alles Likert-Scales von 1-5 bzw. 1-10)

df_float = df.select_dtypes(include='float64')  # nur float-Spalten

df[df_float.columns] = df[df_float.columns].astype("category")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4042 entries, 0 to 4041
Data columns (total 57 columns):
 #   Column                        Non-Null Count  Dtype   
---  ------                        --------------  -----   
 0   submission_id                 4042 non-null   category
 1   age                           4011 non-null   category
 2   cups                          3949 non-null   category
 3   where_drink                   3972 non-null   object  
 4   brew                          3657 non-null   object  
 5   brew_other                    678 non-null    object  
 6   purchase                      710 non-null    object  
 7   purchase_other                31 non-null     object  
 8   favorite                      3980 non-null   category
 9   favorite_specify              114 non-null    object  
 10  additions                     3959 non-null   object  
 11  additions_other               48 non-null     object  
 12  dairy                         1686 non-null   ob

### Kategorische Variablen mit Mehrfachantworten

Variablen, bei denen Mehrfachantworten möglich waren: where_drink, brew, purchase, additions, dairy, sweetener, why_drink

In [16]:
# Spalten mit Mehrfachantworten identifizieren

df_obj = df.select_dtypes(include='object')  # nur object-Spalten
cols = df_obj.columns[~df_obj.columns.str.contains('other|notes|specify')] # ohne Freitext Spalten



In [17]:
# Mehrfachantworten: häufige Kombinationen lassen, ansonsten "other" oder "multiple" o.Ä.

def summarize_mehrfachantwort(df, col, top_n):
    '''
    Adds a new variable to the dataframe that contains the most frequent categories of a column
    and replaces other categories with "other".

    df: pandas dataframe to be used
    col: categorical columns in the dataframe to be transformed
    top_n: integer, the top unique categories that should be retained
    '''
    
    top_list = list(df[col].value_counts(ascending = False).index[:top_n])
    df[f"{col}_top{top_n}"] = df[col]
    msk = df[f"{col}_top{top_n}"].isin(top_list)
    df.loc[~msk, f"{col}_top{top_n}"] = "Other"
    df[f"{col}_top{top_n}"] = df[f"{col}_top{top_n}"].astype("category")

    return df
    
for col in cols: 
    summarize_mehrfachantwort(df, col, 6)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4042 entries, 0 to 4041
Data columns (total 64 columns):
 #   Column                        Non-Null Count  Dtype   
---  ------                        --------------  -----   
 0   submission_id                 4042 non-null   category
 1   age                           4011 non-null   category
 2   cups                          3949 non-null   category
 3   where_drink                   3972 non-null   object  
 4   brew                          3657 non-null   object  
 5   brew_other                    678 non-null    object  
 6   purchase                      710 non-null    object  
 7   purchase_other                31 non-null     object  
 8   favorite                      3980 non-null   category
 9   favorite_specify              114 non-null    object  
 10  additions                     3959 non-null   object  
 11  additions_other               48 non-null     object  
 12  dairy                         1686 non-null   ob

### Kategorien ordnen

In [18]:
# Define the desired order of categories
age_order = ['<18 years old', '18-24 years old', '25-34 years old', '35-44 years old', '45-54 years old', '55-64 years old', '>65 years old']
cups_order = ['Less than 1', '1', '2', '3', '4', 'More than 4']
strength_order = ['Weak', 'Somewhat light', 'Medium', 'Somewhat strong', 'Very strong']
spend_order = ['<$20', '$20-$40', '$40-$60', '$60-$80', '$80-$100', '>$100']
paid_order = ['Less than $2', '$2-$4', '$4-$6', '$6-$8', '$8-$10', '$10-$15', '$15-$20', 'More than $20'] # same for most_paid and most_willing
equipment_order = ['Less than $20', '$20-$50', '$50-$100', '$100-$300', '$300-$500', '$500-$1000', 'More than $1,000']
education_order = ['Less than high school', 'High school graduate', "Some college or associate's degree", "Bachelor's degree", "Master's degree", 'Doctorate or professional degree']

#roast_level? caffeine?

# Apply the category order
df["age"] = df["age"].cat.set_categories(age_order, ordered=True)
df["cups"] = df["cups"].cat.set_categories(cups_order, ordered=True)
df["strength"] = df["strength"].cat.set_categories(strength_order, ordered=True)
df["total_spend"] = df["total_spend"].cat.set_categories(spend_order, ordered=True)
df["most_paid"] = df["most_paid"].cat.set_categories(paid_order, ordered=True)
df["most_willing"] = df["most_willing"].cat.set_categories(paid_order, ordered=True)
df["spent_equipment"] = df["spent_equipment"].cat.set_categories(equipment_order, ordered=True)
df["education_level"] = df["education_level"].cat.set_categories(education_order, ordered=True)


## One-hot encoding

In [19]:
def bool_mehrfachantworten(df, col, seperator):

    # Schritt 1: dummy-encoding
    dummies = df[col].str.get_dummies(sep = seperator)

    # Variablen umbenennen
    dummies = dummies.add_prefix(f"{col} ")

    # Aus 0/1 True/False machen
    dummies = dummies.astype(bool)

    # Schritt 2: mit ursprünglichem Dataframe zusammenführen
    df = pd.concat([df, dummies], axis=1)

    return df

In [20]:
one_hot_cols = ["where_drink", "brew", "additions", "dairy", "sweetener", "why_drink"]

for col in one_hot_cols:
    df = bool_mehrfachantworten(df, col, ", ")

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4042 entries, 0 to 4041
Columns: 110 entries, submission_id to why_drink Other
dtypes: bool(46), category(47), object(17)
memory usage: 1.1+ MB


## Save prepared data

In [21]:
df.to_pickle('coffee_data.pkl')